hub/src/server/utils.ts

```diff
@@ -63,3 +62,36 @@ export function timeout(milliseconds: number): Promise<void> {
     }, milliseconds)
   })
 }
+
+
+export class AsyncMutexScope {
+
+  private readonly _opened = new Set<string>()
+
+  public get openedCount() {
+    return this._opened.size
+  }
+
+  /**
+   * If no mutex of the given `id` is already taken, then a mutex is created and the 
+   * given promise is invoked. The mutex is released once the promise resolves -- either by 
+   * success or error. 
+   * @param id A unique mutex name used in a Map.
+   * @param spawnOwner A function that creates a Promise if the mutex is acquired. 
+   * @returns `true` if the mutex was acquired, otherwise returns `false`
+   */
+  public tryAcquire(id: string, spawnOwner: () => Promise<void>): boolean {
+    if (this._opened.has(id)) {
+      return false
+    }
+
+    // Wrap in Promise.resolve to ensure potential synchronous errors are not throw within this function. 
+    const owner = Promise.resolve().then(() => spawnOwner())
+    this._opened.add(id)
```

Comment:
This looks racey to me.  Is this execution schedule possible?  I.e. can the instantiation of `owner` in one coroutine cause it to yield to another coroutine trying to execute on the same path, such as below?  I suspect the answer will depend on implementation subtleties in node.js?

```
[Request 1]                    |[Request 2]
-------------------------------|-------------------------------------
tryAcquire("/foo")             |
if (this._opened.has(id)) {    |
   return false;               |
}                              |
                               |
const owner = Promise          |
  .resolve()                   |
  .then(() => spawnOwner())    |
                               |
<sleep>                        | <wakeup>
                               | tryAcquire("/foo")
                               | if (this._opened.has(id)) {
                               |   return false;
                               | }
                               |
                               | const owner = Promise
                               |  .resolve()
                               |  .then(() => spawnOwner()
<wakeup>                       | <sleep>
this._opened.add(id)           |
owner.finally(...              |
<sleep>                        | <wakeup>
                               | this._opened.add(id)  /* PROBLEM! */

Reply:
Good catch. The `.then` operation, even if synchronous, adds the inner invocation to the end of the event loop. Updated and double checked with some assertions:

```js
    let test = 1

    const runSpawn = () => {
      test++;
      return spawnOwner()
    }

    const owner = runSpawn()

    assert.equal(test, 2);
    this._opened.add(id)
```

The mutex acquirer was updated to handle potential synchronous exceptions that the `runSpawn()` promise may throw.

Reply:
I don't think this fixes it?  Why is it still this:

```
const owner = spawnOwner()
this._opened.add(id)
```

Instead of this?

```
this._opened.add(id)
const owner = spawnOwner()
```

Reply:
In the latter case, there's _no way_ the `spawnOwner()` call can trigger a coroutine yield.

Reply:
I think that re-ordering could cause a dead lock if `spawnOwner` throws an exception.

Reply:
So we'd want to wrap `spawnOwner()` in a try/catch/finally block where the `finally` branch de-registers `id`?

Reply:
The original race condition was because `spawnOwner()` was invoked within a `.then(...)` callback which caused it to execute _after_ the proceeding `this._opened.add(id)` statement due to event-loop behavior of promise chaining.

With the current code, the `spawnOwner()` statement is guaranteed to execute all of its _synchronous_ code before the `this._opened.add(id)` statement executes. 

However, there is a potential problem if callers attempt reentrant locking. That can be fixed by re-order the statements. The downside is having to re-throw the error within the catch statement after de-registering.

Reply:
https://github.com/blockstack/gaia/pull/247/commits/a7f04dad4d182310a927999b49f1997d5af1ab54 fixes the potential synchronous reentrancy issue if this class gets misused later. Uses a try/catch/re-throw to deal with synchronous errors.

Reply:
LGTM